rh-hideout/pokeemerald-expansion
# Add B_LEVEL_UP_NOTIFICATION to improve player QoL when performing multiple level ups

# Description
![Gif of Pikipek getting enough experience to jump from level 1 to 20](https://files.catbox.moe/k32wui.gif)

This PR adds `B_LEVEL_UP_NOTIFICATION` to `include/config/battle.h`. When set to `GEN_9` or later, when Pokemon receive enough EXP in battle to progress through multiple levels, the level up message is only shown once.

## Details

### Usage
In [`include/config/battle.h`](https://github.com/PokemonSanFran/pokeemerald-expansion/blob/levelUpNotification/include/config/battle.h), developers must define which generation `B_LEVEL_UP_NOTIFICATION` should use.

## => `GEN_9`
When a Pokemon receives enough experience to grow multiple levels, the following changes occur: 
* The message "MON grew to LEVEL X" only shows once.
* All moves to be learn are attempted to be taught after the level up Notification.
* If the mon is currently participating in battle, the EXP bar only animates once.
* After battle, if the Pokemon is eligible, it will try to evolve.

## < `GEN_9`
Any value lower than `GEN_9` will handle leveling up the same as vanilla.

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://files.catbox.moe/1uugqs.patch ; git apply 1uugqs.patch ; rm 1uugqs.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion levelUpNotification`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script, or manually create the changes.

### Download
* [`data/scripts/debug.inc`](https://github.com/PokemonSanFran/pokeemerald-expansion/blob/levelUpNotification_testing/data/scripts/debug.inc)

### Manual Testing
* Compile the game
* Start a new save
* Run Script 1
* Run Script 2
* Switch to Tyranitar
* Observe functionality

## Verified Scenarios

Both videos show:
* Player starts a battle with a wild Level 100 Shedinja versus their Level 1 Pikipek
* Player switches to a Tyranitar
* Shedinja faints from sandstorm damage and Pikipek grows to level 20
* Pikipek learns Growl, Echoed Voice, Rock Smash.
* Pikipek is prompted to, but does not learn Supersonic, Pluck, and Roost.
* Pikipek evolves into Trumbeak.
* Player starts a battle with a wild Level 100 Shedinja versus their Level 20 Trumbeak
* Player uses debug menu to give Trumbreak 9044 ATK and 9038 SPEED
* Shedinja faints from Trumbeak's Peck and Trumbeak grows to Level 23
* Trumbeak is prompted to, but does not learn Roost.

### `B_LEVEL_UP_NOTIFICATION == GEN_LATEST`

https://github.com/user-attachments/assets/3210b5db-af7e-4432-aa0b-542ee5f2ed7a

### `B_LEVEL_UP_NOTIFICATION == GEN_8`

https://github.com/user-attachments/assets/08df6b0e-f111-4b70-bca0-62b87b5680e7

# **People who collaborated with me in this PR**
@DizzyEggg wrote all of this.

# Features this PR does NOT handle:
- This PR **DOES NOT** handle EXP gained outside of battle.

# Discord Contact Info

I am `pkmnsnfrn` on Discord.
